### ISSUE TITLE
HiLo sequence incorrectly shared between different database connections

### ISSUE THREAD

**Issue by Systemdir** [aspnet/EntityFrameworkCore#11451]
We noticed that one tests fail when executed together with another test. Narrowing it down we found out that the reason is npgsql HiLo when using the same context to connect to different databases on the same database server (we have this setup to be able to run multiple CI tests at the same time).

### Steps to reproduce
**HiLo Setup:**
```C#
public static void AddHiLoSequence<T>(this ModelBuilder modelBuilder, Expression<Func<T, long>> property) where T : class
{
	string sequenceName = typeof(T).Name+"HiLoSequence";

	modelBuilder.HasSequence<long>(sequenceName)
		.StartsAt(1)
		.IncrementsBy(10);

	modelBuilder.Entity<T>()
		.Property(property)
		.ForNpgsqlUseSequenceHiLo(sequenceName);
}
````

**DatabaseFixture (added it for complenteness sake):**
```C#
public DatabaseFixture()
{
	_databaseNameSuffix = Guid.NewGuid();
	
        // get connection strings
	string dbServer = Environment.GetEnvironmentVariable("dbServer") ?? "localhost";
	string dbPort = Environment.GetEnvironmentVariable("dbPort") ?? "5432";

	_connectionString =
		$@"Server={dbServer};Port={dbPort};Database=atmentis{
				_databaseNameSuffix.Value
			};User Id=postgres;Password=...";
	_readOnlyConnectionString =
		$@"Server={dbServer};Port={dbPort};Database=atmentis{
				_databaseNameSuffix.Value
			};User Id=postgres;Password=...";

	// create database structure
	CreateDbContext().Database.EnsureCreated();
}
```
**Failing Test:**

```C#
// getting a random database 
// (e.g. Server=localhost;Port=5432;Database=atmentis7156f551-b8a8-4b0f-a3d2-a6dc5e219ea9;User Id=postgres;Password=...)
DatabaseFixture databaseFixture1 = new DatabaseFixture();

using (AtmentisContext atmentisContext1 = databaseFixture1.CreateDbContext())
{
	using (IDbContextTransaction transaction1 = atmentisContext1.Database.BeginTransaction())
	{
		ClipDatabase clipDatabase = new ClipDatabase()
		{
			ContentProvider = ContentProvider.Youtube,
			Title = "clip context 1",
			Url = ""
		};
		atmentisContext1.Clips.Add(clipDatabase);
		atmentisContext1.SaveChanges();
		transaction1.Commit();
	}
}

// getting another random database connection 
// (e.g.  Server=localhost;Port=5432;Database=atmentis1e8c9f5e-3907-4508-b49f-869f67189b4e;User Id=postgres;Password=...")
DatabaseFixture databaseFixture2 = new DatabaseFixture();

for (int i = 0; i < 15; i++)
{
	using (AtmentisContext atmentisContext2 = databaseFixture2.CreateDbContext())
	{
		using (IDbContextTransaction transaction2 = atmentisContext2.Database.BeginTransaction())
		{
			ClipDatabase clipDatabase = new ClipDatabase()
			{
				ContentProvider = ContentProvider.Youtube,
				Title = "clip context 2",
				Url = ""
			};
			atmentisContext2.Clips.Add(clipDatabase);
			atmentisContext2.SaveChanges();
			transaction2.Commit();
		}
	}
}

```

### The issue
**SQL Log for the first database connection:**
```SQL
2018-03-26 15:00:25.218 UTC [742] LOG:  statement: DISCARD ALL
2018-03-26 15:00:25.218 UTC [742] LOG:  statement: BEGIN
2018-03-26 15:00:25.218 UTC [742] LOG:  statement: SET TRANSACTION ISOLATION LEVEL READ COMMITTED
2018-03-26 15:00:25.219 UTC [742] LOG:  execute <unnamed>: SELECT nextval('"ClipDatabaseHiLoSequence"')
2018-03-26 15:00:25.333 UTC [742] LOG:  execute <unnamed>: INSERT INTO "ContentLibrary"."Clip" ("Id", "ContentProvider", "Title", "Url")
        VALUES ($1, $2, $3, $4)
2018-03-26 15:00:25.333 UTC [742] DETAIL:  parameters: $1 = '1', $2 = 'Youtube', $3 = 'clip context 1', $4 = ''
2018-03-26 15:00:25.360 UTC [742] LOG:  statement: COMMIT
```

**SQL Log for the second database connection:**
```SQL
# The next commands fail to get HiLo Sequence next value (which would start by 1).
# It continues using the value of the other db connection and inserts with id 2.

2018-03-26 15:01:11.994 UTC [781] LOG:  statement: DISCARD ALL
2018-03-26 15:01:11.994 UTC [781] LOG:  statement: BEGIN
2018-03-26 15:01:11.994 UTC [781] LOG:  statement: SET TRANSACTION ISOLATION LEVEL READ COMMITTED
2018-03-26 15:01:11.994 UTC [781] LOG:  execute <unnamed>: INSERT INTO "ContentLibrary"."Clip" ("Id", "ContentProvider", "Title", "Url")
        VALUES ($1, $2, $3, $4)
2018-03-26 15:01:11.994 UTC [781] DETAIL:  parameters: $1 = '2', $2 = 'Youtube', $3 = 'clip context 2', $4 = ''
2018-03-26 15:01:12.007 UTC [781] LOG:  statement: COMMIT


2018-03-26 15:01:32.485 UTC [781] LOG:  statement: DISCARD ALL
2018-03-26 15:01:32.485 UTC [781] LOG:  statement: BEGIN
2018-03-26 15:01:32.485 UTC [781] LOG:  statement: SET TRANSACTION ISOLATION LEVEL READ COMMITTED
2018-03-26 15:01:32.485 UTC [781] LOG:  execute <unnamed>: INSERT INTO "ContentLibrary"."Clip" ("Id", "ContentProvider", "Title", "Url")
        VALUES ($1, $2, $3, $4)
2018-03-26 15:01:32.485 UTC [781] DETAIL:  parameters: $1 = '3', $2 = 'Youtube', $3 = 'clip context 2', $4 = ''
2018-03-26 15:01:32.516 UTC [781] LOG:  statement: COMMIT

# ... 

2018-03-26 15:02:03.379 UTC [781] LOG:  statement: DISCARD ALL
2018-03-26 15:02:03.380 UTC [781] LOG:  statement: BEGIN
2018-03-26 15:02:03.380 UTC [781] LOG:  statement: SET TRANSACTION ISOLATION LEVEL READ COMMITTED
2018-03-26 15:02:03.380 UTC [781] LOG:  execute <unnamed>: INSERT INTO "ContentLibrary"."Clip" ("Id", "ContentProvider", "Title", "Url")
        VALUES ($1, $2, $3, $4)
2018-03-26 15:02:03.380 UTC [781] DETAIL:  parameters: $1 = '10', $2 = 'Youtube', $3 = 'clip context 2', $4 = ''
2018-03-26 15:02:03.388 UTC [781] LOG:  statement: COMMIT

# The next commands get the next HiLo Sequence value which is 1 as it failed to get nextval before
# As we started at id 2 above, the command succeeds

2018-03-26 15:02:32.329 UTC [781] LOG:  statement: DISCARD ALL
2018-03-26 15:02:32.329 UTC [781] LOG:  statement: BEGIN
2018-03-26 15:02:32.329 UTC [781] LOG:  statement: SET TRANSACTION ISOLATION LEVEL READ COMMITTED
2018-03-26 15:02:32.330 UTC [781] LOG:  execute <unnamed>: SELECT nextval('"ClipDatabaseHiLoSequence"')
2018-03-26 15:02:32.332 UTC [781] LOG:  execute <unnamed>: INSERT INTO "ContentLibrary"."Clip" ("Id", "ContentProvider", "Title", "Url")
        VALUES ($1, $2, $3, $4)
2018-03-26 15:02:32.332 UTC [781] DETAIL:  parameters: $1 = '1', $2 = 'Youtube', $3 = 'clip context 2', $4 = ''
2018-03-26 15:02:32.338 UTC [781] LOG:  statement: COMMIT

# The next commands will get the next sequence value which is 2 and fail as id 2 was already inserted (see above)

2018-03-26 15:02:53.754 UTC [781] LOG:  statement: DISCARD ALL
2018-03-26 15:02:53.754 UTC [781] LOG:  statement: BEGIN
2018-03-26 15:02:53.754 UTC [781] LOG:  statement: SET TRANSACTION ISOLATION LEVEL READ COMMITTED
2018-03-26 15:02:53.754 UTC [781] LOG:  execute <unnamed>: INSERT INTO "ContentLibrary"."Clip" ("Id", "ContentProvider", "Title", "Url")
        VALUES ($1, $2, $3, $4)
2018-03-26 15:02:53.754 UTC [781] DETAIL:  parameters: $1 = '2', $2 = 'Youtube', $3 = 'clip context 2', $4 = ''
2018-03-26 15:02:53.754 UTC [781] ERROR:  duplicate key value violates unique constraint "PK_Clip"
2018-03-26 15:02:53.754 UTC [781] DETAIL:  Key ("Id")=(2) already exists.
2018-03-26 15:02:53.754 UTC [781] STATEMENT:  INSERT INTO "ContentLibrary"."Clip" ("Id", "ContentProvider", "Title", "Url")
        VALUES ($1, $2, $3, $4)
2018-03-26 15:02:53.861 UTC [781] LOG:  statement: ROLLBACK

```

**Exception:**
```C#
Microsoft.EntityFrameworkCore.DbUpdateException : An error occurred while updating the entries. See the inner exception for details.
---- Npgsql.PostgresException : 23505: duplicate key value violates unique constraint "PK_Clip"
   at Microsoft.EntityFrameworkCore.Update.ReaderModificationCommandBatch.Execute(IRelationalConnection connection)
   at ...

```

### Further technical details
EnityFramework version: Microsoft.AspNetCore.Identity.EntityFrameworkCore v2.0.2
Npgsql version: Npgsql.EntityFrameworkCore.PostgreSQL v2.0.1
PostgreSQL version: PostgreSQL 10.3 (Debian 10.3-1.pgdg90+1) on x86_64-pc-linux-gnu, compiled by gcc (Debian 6.3.0-18+deb9u1) 6.3.0 20170516, 64-bit
Operating system: (Docker) Debian GNU/Linux 9

[Issue moved from npgsql/Npgsql.EntityFrameworkCore.PostgreSQL#350]

**Comment by Fmk-Milestone** [comment]
I am working with EF core 2.2 and dynamically setting connection string for each DbContext.

The above problem is occurring in my application. Do I need to set any special setting for DBContext.

**Scenario :**

Connection string is generated on the basis of user.

User A  has Db1
User B  has Db2

User A belongs to Company A
User B belongs to Company B

When User A fetches sequence from Db1 it is shared with User B which is wrong.

As sequence is not cache per database wise

Any help will be appreciated.

**Comment by ajcvickers** [comment]
@Fmk-Milestone EF Core 2.2 is now [out-of-support](https://docs.microsoft.com/en-us/ef/core/what-is-new/). This issue was fixed in EF Coere 3.0, as indicated by the milestone set and the closed-fixed label. Please update to EF Core 3.1 and let us know if you're still running into issues with this.

**Comment by mandm-pt** [comment]
Any workaround for people using 2.2? I am using Oracle which depends on 2.2, and don't know when I will get a package for 3.1

**Comment by davidroth** [comment]
For reference, this has only been fixed in ef-core 3.0.0 but not in ef-core postgres. So this issue still occurs when using the postgres provider.

See issue: https://github.com/npgsql/efcore.pg/issues/1394
